packages/lesswrong/lib/tableOfContents.ts

```diff
@@ -28,5 +34,251 @@ export interface ToCSectionWithOffset extends ToCSection {
 export interface ToCData {
   html: string | null,
   sections: ToCSection[],
-  headingsCount: number,
+}
+
+// Number of headings below which a table of contents won't be generated.
+// If comments-ToC is enabled, this is 0 because we need a post-ToC (even if
+// it's empty) to keep the horizontal position of things on the page from
+// being imbalanced.
+export const MIN_HEADINGS_FOR_TOC = commentsTableOfContentsEnabled ? 0 : 1;
+
+// Tags which define headings. Currently <h1>-<h4>, <strong>, and <b>. Excludes
+// <h5> and <h6> because their usage in historical (HTML) wasn't as a ToC-
+// worthy heading.
+const headingTags = {
+  h1: 1,
+  h2: 2,
+  h3: 3,
+  h4: 4,
+  // <b> and <strong> are at the same level
+  strong: 7,
+  b: 7,
+}
+
+const headingIfWholeParagraph = {
+  strong: true,
+  b: true,
+};
+
+const headingSelector = _.keys(headingTags).join(",");
+
+/**
+ * Given an HTML document, extract a list of sections for a table of contents
+ * from it, and add anchors. The result is modified HTML with added anchors,
+ * plus a JSON array of sections, where each section has a
+ * `title`, `anchor`, and `level`, like this:
+ *   {
+ *     html: "<a anchor=...">,
+ *     sections: [
+ *       {title: "Preamble", anchor: "preamble", level: 1},
+ *       {title: "My Cool Idea", anchor: "mycoolidea", level: 1},
+ *         {title: "An Aspect of My Cool Idea", anchor:"anaspectofmycoolidea", level: 2},
+ *         {title: "Why This Is Neat", anchor:"whythisisneat", level: 2},
+ *       {title: "Conclusion", anchor: "conclusion", level: 1},
+ *     ]
+ *   }
+ */
+export function extractTableOfContents({
+  document,
+  window,
+}: {
+  document: Document;
+  window: WindowType;
+}): ToCData | null {
+  if (!document.body.innerHTML) return null;
+
+  let headings: Array<ToCSection> = [];
+  let usedAnchors: Record<string, boolean> = {};
+
+  // First, find the headings in the document, create a linear list of them,
+  // and insert anchors at each one.
+  let headingElements = document.querySelectorAll(headingSelector);
+  headingElements.forEach((element) => {
```

Comment:
`forEach` considered harmful, please use `for ... of` [instead](https://github.com/ForumMagnum/ForumMagnum/pull/5505#discussion_r953136414).

Reply:
And elsewhere